### DISCUSSION TITLE
How to automatically turn esprima AST into observable

### DISCUSSION THREAD

**Discussion by brabeji** [mobxjs/mobx#3124] [Q&A] [answered]
Hi guys, looking for some help. I need to (preferably automatically) turn AST produced by esprima parser into an observable tree. The problem is that the parser produces class instances that are not under my control. Therefore the tree is not automatically transformed to observable as I would like.

I could walk the tree and maybe call `extendObservable` for each node but that would mean handling each node type (~100) separately. That I would like to avoid.

Also, I'm stuck at mobx@5 for now. Thank you for any suggestions.

**Comment by brabeji** [comment, marked as answer]
Solved by cloning the AST while omitting constructors.

**Comment by mweststrate** [comment]
Wondering if calling `makeAutoObservable(instance)` on every node
recursively would do the trick as well?

**Comment by mweststrate** [comment]
Sorry, that exists only on 6. reflection + extendObservable for every own
property would probably do the trick on 5.

**Reply by brabeji** [reply]
reflection + extendObservable was also my thought. The current code is less complex at the expense of cloning a rather small tree 😅  will see how that works out performance-wise 🤷‍♂️  Thanks!
